### PR TITLE
Add MiniRL Support

### PR DESCRIPTION
**Pull request by shamanez**
### What does this PR do?

This PR adds an initial implementation of the MINRL objective described in  
[arXiv:2512.01374](https://arxiv.org/pdf/2512.01374).

Concretely, it introduces a new loss helper `compute_policy_loss_minirl` in
`core_algos.py`, which is intended to be used as an alternative policy loss
for future experiments.

This is still work-in-progress: the implementation compiles and runs on small
tests, but I have not yet trained a full model due to limited compute.
Feedback and suggestions on the design and integration are very welcome.

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here:  
  https://github.com/volcengine/verl/pulls?q=MINRL
- [x] Format the PR title as `[{modules}] {type}: {description}`  
  (using `[algo] feat: add MINRL policy loss`)

### Test

I do not have enough compute to run a full training run yet.

Current checks:
- [x] Code passes local linting/formatting.
- [x] Basic smoke test on a small batch to ensure `compute_policy_loss_minirl`
      runs without errors and returns finite losses.

Planned:
- [ ] Full training run once compute is available.
- [ ] Compare learning curves against the default policy loss.

### CODE REVIEW

File: verl/trainer/ppo/core_algos.py [hunk 1]
```diff
@@ -1073,6 +1073,159 @@ def compute_policy_loss_gspo(
...
+    #       L = - M_t ⋅ w_t(θ) ⋅ Â(x,y) ⋅ log π_θ
+    #
+    #    Gradients:
+    #       ∇_θ L = - E[ M_t ⋅ w_t(θ) ⋅ Â ⋅ ∇ log π_θ ]
+    #
+    #    where:
+    #       - M_t only gates tokens (built from r_detached)
+    #       - w_t(θ) combines staleness + rollout mismatch
+    # -------------------------------------------------------------------------
+    pg_losses = -gate * total_is_weight * advantages * log_prob
```
[the remark this answers is not in the document]
Comment (shamanez) [reply verl/trainer/ppo/core_algos.py:1198]: According to the attached paper, this loss function is more of a REINFORCE like objective right ? Then the ratio is only being used for a decoupled clip.
[QWEN-lessons.pdf](https://github.com/user-attachments/files/24097810/QWEN-lessons.pdf)

[the remark this answers is not in the document]
Comment (xiatian-wuhan) [reply verl/trainer/ppo/core_algos.py]: def compute_token_level_is_weights(
        self,
        log_probs_rollout_train: torch.Tensor,  # Shape: (batch_size, seq_len)
        log_probs_rollout_infer: torch.Tensor,  # Shape: (batch_size, seq_len)
        attention_mask: torch.Tensor,            # Shape: (batch_size, seq_len)
    ) -> torch.Tensor:
        """
        Compute token-level importance sampling weights

        w_t = π_θ_old^train(y_t|x,y_<t) / μ_θ_old^infer(y_t|x,y_<t)

        This corrects for the training-inference discrepancy.

        Args:
            log_probs_rollout_train: Log probs from training engine during rollout
            log_probs_rollout_infer: Log probs from inference engine during rollout
            attention_mask: Mask for valid tokens (1 for valid, 0 for padding)

        Returns:
            is_weights: Token-level IS weights (gradient stopped)
        """
        if not self.config.use_train_infer_is:
            # Return ones if not using IS correction
            return torch.ones_like(log_probs_rollout_train)

        # Compute log ratio
        log_ratio = log_probs_rollout_train - log_probs_rollout_infer

        # Compute ratio (with clipping for numerical stability)
        ratio = torch.exp(torch.clamp(log_ratio, min=-20, max=20))

        # Apply truncated IS if threshold is set
        if self.config.tis_threshold is not None:
            ratio = torch.clamp(ratio, max=self.config.tis_threshold)

        # Apply attention mask
        ratio = ratio * attention_mask

        # Stop gradient (sg operation)
        ratio = ratio.detach()

        return ratio

**Comment by shamanez** [comment]
Some convergence proof.

Screenshot 2025-12-17 at 11 39 46 am
Screenshot 2025-12-17 at 11 41 52 am
Screenshot 2025-12-17 at 11 42 21 am
Screenshot 2025-12-17 at 11 43 38 am

@eric-haibin-lin @vermouth1992

**Comment by szrlee** [comment]
@shamanez please take a deep look at the rollout correction module in VeRL. https://verl.readthedocs.io/en/latest/algo/rollout_corr_math.html
You may reuse this module.

**Comment by shamanez** [comment]
@szrlee  Hello Yingru, I actually used the rollout correction module, but  according to this paper, we also need to compute staleness ratio, which is the ratio between pi_old and the pi_theta (what we update during the ppo mini batch.)

That is why I computed that inside the loss function.

**Comment by shamanez** [comment]
@szrlee I actually reused the error correction module, and explained the logic inside the loss dunction. Please let me know if there's anything missing .

**Comment by szrlee** [comment]
> @szrlee I actually reused the error correction module, and explained the logic inside the loss dunction. Please let me know if there's anything missing .

@shamanez what is the difference with decoupled_token_is() in https://verl.readthedocs.io/en/latest/algo/rollout_corr_math.html#method-summary-table?

using token_is at \pi_{rollout} -> \pi_{old} and ppo_{clip} for \pi_{old} -> \pi_{new}?

There is no need to implement another loss for miniRL accounting for \pi_{old} -> \pi_{new} as it is essentially ppo_clip loss. You can either check the math definition or code or Qwen's paper. It is just rebranding.

**Comment by shamanez** [comment]
@szrlee ok I understand, the` Bypass Mode Presets (PPO-clip)` does almost the same.  Does it mean, this loss function in the core_algo.py (compute_policy_loss_with_rollout_correction) , implements the logic ? 

But could you please help me to understand if the current PPO clip loss method also addresses following constrains. 

  Well please correct me if I'm am wrong, but according to their logic ,  the importance sampling weight depends on two ratios as follows. 

image

If I understand it correctly, here pi_theta is actually the loss getting updated inside the mini_batch ppo. In the error correction module only compute the training-internee discrepancy once per the batch.

Also they use decoupled PPO.
image

The reason, I thought of adding it as a new loss is , as you can see in their experiments, pi_theta is basically the policy that is getting updated in the mini batch..

image

**Comment by szrlee** [comment]
> @szrlee ok I understand, the` Bypass Mode Presets (PPO-clip)` does almost the same. Does it mean, this loss function in the core_algo.py (compute_policy_loss_with_rollout_correction) , implements the logic ?
> 
> But could you please help me to understand if the current PPO clip loss method also addresses following constrains.
> 
> Well please correct me if I'm am wrong, but according to their logic , the importance sampling weight depends on two ratios as follows.
> 
> image
> If I understand it correctly, here pi_theta is actually the loss getting updated inside the mini_batch ppo. In the error correction module only compute the training-internee discrepancy once per the batch.
> 
> Also they use decoupled PPO. image
> 
> The reason, I thought of adding it as a new loss is , as you can see in their experiments, pi_theta is basically the policy that is getting updated in the mini batch..
> 
> image

@shamanez  Please look at the preset methods on deoupled_token_is() in the algorithm.py and see how it works. thank you!

We already address both 3 policy case (\pi_{rollout} -> \pi_{old}, \pi_{old} -> \pi_{new}) and 2 policy case (\pi_{rollout} -> \pi_new) in rollout correction module.

Qwen's paper also cites our work.
